File: backend/src/middlewares/auth.middleware.ts
import { Request, Response, NextFunction } from 'express';
import { TokenUtils } from '../utils/token.utils';
import db from '../database/db';
import { User } from '../interfaces/user.interface';

export interface AuthRequest extends Request {
  user?: User;
}

export async function authMiddleware(
  req: AuthRequest,
  res: Response,
  next: NextFunction
): Promise<void> {
  try {
    const authHeader = req.headers.authorization;

    if (!authHeader || !authHeader.startsWith('Bearer ')) {
      res.status(401).json({ detail: 'Not authenticated' });
      return;
    }

    const token = authHeader.replace('Bearer ', '');
    const payload = TokenUtils.verifyAccessToken(token);

    if (!payload) {
      res.status(401).json({ detail: 'Invalid or expired token' });
      return;
    }

    const user = await db<User>('users')
      .where({ id: payload.sub, deleted_at: null })
      .first();

    if (!user) {
      res.status(401).json({ detail: 'User not found' });
      return;
    }

    req.user = user;
    next();
  } catch (error) {
    res.status(401).json({ detail: 'Invalid token' });
  }
}

File: backend/middleware/auth_middleware.py
from fastapi import HTTPException, Header, Depends
from typing import Optional

from utils.token_utils import decode_access_token
from dependencies import get_database


async def get_current_user(authorization: Optional[str] = Header(None), db=Depends(get_database)):
    """Middleware to extract and validate current user from token"""
    if not authorization or not authorization.startswith("Bearer "):
        raise HTTPException(status_code=401, detail="Not authenticated")
    
    token = authorization.replace("Bearer ", "")
    payload = decode_access_token(token)
    
    if not payload:
        raise HTTPException(status_code=401, detail="Invalid or expired token")
    
    user_id = payload.get("sub")
    if not user_id:
        raise HTTPException(status_code=401, detail="Invalid token payload")
    
    user = await db.users.find_one(
        {"id": user_id, "deleted_at": None},
        {"_id": 0}
    )
    
    if not user:
        raise HTTPException(status_code=401, detail="User not found")
    
    return user
